### PR TITLE
Clear terminal output buffers to prevent memory leaks

### PR DESCRIPTION
**Pull request by hannesrudolph**
## Summary
- avoid retaining large command output by trimming buffers in BaseTerminalProcess
- clear stored output when TerminalProcess and ExecaTerminalProcess finish running commands

## Testing
- `pnpm exec vitest run integrations/terminal/__tests__/TerminalProcess.spec.ts integrations/terminal/__tests__/ExecaTerminalProcess.spec.ts`
- `pnpm exec eslint integrations/terminal/BaseTerminalProcess.ts integrations/terminal/ExecaTerminalProcess.ts integrations/terminal/TerminalProcess.ts`

------
https://chatgpt.com/codex/tasks/task_e_68b9a2189e148330b3b21e8c713050d6

### CODE REVIEW

File: src/integrations/terminal/BaseTerminalProcess.ts [hunk 1]
```diff
@@ -137,6 +137,20 @@ export abstract class BaseTerminalProcess extends EventEmitter<RooTerminalProces
...
 
+	/**
+	 * Trims already retrieved content from the internal output buffer.
+	 *
+	 * This prevents unbounded memory growth when processing large
+	 * command outputs by discarding data that has already been
+	 * consumed by callers of `getUnretrievedOutput`.
+	 */
+	protected trimRetrievedOutput(): void {
+		if (this.lastRetrievedIndex > 0) {
```
[the remark this answers is not in the document]
Comment (hannesrudolph) [reply src/integrations/terminal/BaseTerminalProcess.ts]: Good catch! I've applied your suggested fix. The condition is now `lastRetrievedIndex >= fullOutput.length && fullOutput.length > 0` which ensures the buffer is only cleared when all output has been retrieved and the buffer isn't already empty. This was addressed in commit 1c658491f5.

File: src/integrations/terminal/ExecaTerminalProcess.ts [hunk 2]
```diff
@@ -146,6 +146,8 @@ export class ExecaTerminalProcess extends BaseTerminalProcess {
 		this.emitRemainingBufferIfListening()
 		this.stopHotTimer()
 		this.emit("completed", this.fullOutput)
+		this.lastRetrievedIndex = this.fullOutput.length
```
[the remark this answers is not in the document]
Comment (hannesrudolph) [reply src/integrations/terminal/ExecaTerminalProcess.ts]: You're right about the race condition. I've moved the buffer cleanup (setting `lastRetrievedIndex` and calling `trimRetrievedOutput()`) to after the "continue" event is emitted. This ensures all event listeners have a chance to access the output if needed. The code now emits "completed", then "continue", and only then clears the buffer. Fixed in commit 1c658491f5.

File: src/integrations/terminal/TerminalProcess.ts [hunk 3]
```diff
@@ -254,6 +254,8 @@ export class TerminalProcess extends BaseTerminalProcess {
 		// so that api request stalls to let diagnostics catch up").
 		this.stopHotTimer()
 		this.emit("completed", this.removeEscapeSequences(this.fullOutput))
+		this.lastRetrievedIndex = this.fullOutput.length
```
[the remark this answers is not in the document]
Comment (hannesrudolph) [reply src/integrations/terminal/TerminalProcess.ts]: Fixed the same race condition here in TerminalProcess. The buffer cleanup now happens after both "completed" and "continue" events are emitted, ensuring consistency with ExecaTerminalProcess and giving all listeners the opportunity to access output before it's cleared. Fixed in commit 1c658491f5.

File: src/integrations/terminal/BaseTerminalProcess.ts [hunk 4]
```diff
@@ -137,6 +137,20 @@ export abstract class BaseTerminalProcess extends EventEmitter<RooTerminalProces
...
 	abstract getUnretrievedOutput(): string
 
+	/**
+	 * Trims already retrieved content from the internal output buffer.
+	 *
+	 * This prevents unbounded memory growth when processing large
+	 * command outputs by discarding data that has already been
+	 * consumed by callers of `getUnretrievedOutput`.
+	 */
+	protected trimRetrievedOutput(): void {
```
[the remark this answers is not in the document]
Comment (hannesrudolph) [reply src/integrations/terminal/BaseTerminalProcess.ts]: Done! I've added test coverage for `trimRetrievedOutput()` to both `TerminalProcess.spec.ts` and `ExecaTerminalProcess.spec.ts`. The tests cover:

1. Buffer is cleared when all output has been retrieved (`lastRetrievedIndex >= fullOutput.length`)
2. Buffer is NOT cleared when there's still unretrieved output (preserves ongoing retrieval)
3. Empty buffer edge case (no-op when buffer is already empty)
4. Edge case where `lastRetrievedIndex` exceeds `fullOutput.length`

All 25 tests pass including the 8 new ones. Added in commit 1c658491f5.

**Comment by daniel-lxs** [comment]
Hey! I traced through the code and I think there's a timing issue here.

The problem is this line:
```typescript
this.lastRetrievedIndex = this.fullOutput.length
```

This marks all output as "retrieved" immediately after the command completes, but `getEnvironmentDetails()` hasn't had a chance to actually call `getUnretrievedOutput()` yet.

What happens:
1. Command completes → process gets added to `completedProcesses` queue ✓
2. Buffer gets cleared (PR change)
3. Later, `getEnvironmentDetails()` tries to get output from completed processes
4. `cleanCompletedProcessQueue()` removes the process because `hasUnretrievedOutput()` now returns false
5. The AI never sees the terminal output

This breaks the "Inactive Terminals with Completed Process Output" section in `getEnvironmentDetails.ts`.

The fix would be to not force `lastRetrievedIndex = fullOutput.length` - just call `trimRetrievedOutput()` by itself, which will only clear when output has actually been consumed via `getUnretrievedOutput()`.

**Comment by hannesrudolph** [comment]
> Hey! I traced through the code and I think there's a timing issue here.
> 
> The problem is this line:
> 
> ```ts
> this.lastRetrievedIndex = this.fullOutput.length
> ```
> 
> This marks all output as "retrieved" immediately after the command completes, but `getEnvironmentDetails()` hasn't had a chance to actually call `getUnretrievedOutput()` yet.
> 
> What happens:
> 
> 1. Command completes → process gets added to `completedProcesses` queue ✓
> 2. Buffer gets cleared (PR change)
> 3. Later, `getEnvironmentDetails()` tries to get output from completed processes
> 4. `cleanCompletedProcessQueue()` removes the process because `hasUnretrievedOutput()` now returns false
> 5. The AI never sees the terminal output
> 
> This breaks the "Inactive Terminals with Completed Process Output" section in `getEnvironmentDetails.ts`.
> 
> The fix would be to not force `lastRetrievedIndex = fullOutput.length` - just call `trimRetrievedOutput()` by itself, which will only clear when output has actually been consumed via `getUnretrievedOutput()`.

@roomote address this please.
